src/arithmetic/arithmetic_expression.c

```diff
@@ -167,15 +175,17 @@ bool AR_EXP_ReduceToScalar(AR_ExpNode **root) {
 	}
 }
 
-static bool _AR_EXP_ValidateInvocation(AR_FuncDesc *fdesc, SIValue *argv, uint argc, char **error) {
+static bool _AR_EXP_ValidateInvocation(AR_FuncDesc *fdesc, SIValue *argv, uint argc) {
 	SIType actual_type;
 	SIType expected_type = T_NULL;
 
 	// Make sure number of arguments is as expected.
 	if(fdesc->argc != VAR_ARG_LEN) {
 		if(fdesc->argc != argc) {
-			asprintf(error, "Received %d arguments to function '%s', expected %d", argc, fdesc->name,
+			char *error;
+			asprintf(&error, "Received %d arguments to function '%s', expected %d", argc, fdesc->name,
 					 fdesc->argc);
+			QueryCtx_SetError(error); // Set the query-level error.
```

Comment:
Would it be nicer if `_RaiseException` would also set the error, seems like the right design to me.
if so `_AR_EXP_ValidateInvocation` would need to return `error`

Reply:
I understand that desire, but there are 8 different places that we can generate an error message, including actual arithmetic operations like `AR_GT`, and we need to be able to unwind to free memory cleanly. I think that this is a cleaner system than propagating back an `SI_Error` and polluting our SIType system as a result.